### PR TITLE
Material.add_material(...) method

### PR DESCRIPTION
**Pull request by samuelshaner**
This short PR adds an `add_material(...)` method to the Material object in the Python API. This method is useful when trying to homogenize multiple materials by atom percent (e.g. for the C5G7 benchmark or water/structural materials in BEAVRS) or weight percent (e.g. for the TREAT fuel/graphite blocks). The `add_material(...)` is structured such that it expands a given material into its constituent nuclides and adds the nuclides to the base material. 

While this method is handy, there are some caveats that haven't been addressed in this PR. Since the temperature is set by material or cell, a material can only be added to another material if they have the same temperature. Also, since s(a,b) are set on a material basis, there is the possibility of a material having multiple s(a,b) for the same nuclide. I'm not sure how OpenMC treats these cases, but it doesn't appear that an error is given.

Two materials can be homogenized as follows:

    fuel = openmc.Material()
    fuel.set_density('g/cm3', 10.0)
    fuel.add_element('U', 1.0, enrichment=2.5)
    fuel.add_element('O', 2.0)

    clad = openmc.Material()
    clad.set_density('g/cm3', 6.55)
    clad.add_element('Zr', 1.0)
 
    fuel_clad = openmc.Material()
    fuel_clad.add_material(fuel, 0.9, 'ao')
    fuel_clad.add_material(clad, 0.1, 'ao')

This PR builds on PR #771

### CODE REVIEW

**Comment by nelsonag** [comment]
A few initial thoughts (though don't wait on me for a full review, I won't have my computer with me for a week starting tomorrow):
- It would be good to check for duplicate S(a,b) data and not add it twice.
- It doesn't look like this combines unique nuclides. That is, if you are combining H2O and H3BO3, then you will have two H and two O entries instead of increasing the fractions of the existing H and O.  Is that a fair assessment?  If so, we should agree on what we want the behavior to be for deaing with multiple nuclides, and (as you point out), how one could assign an S(a,b) to some of the atoms but not all.

**Comment by samuelshaner** [comment]
When making this PR my inclination was to give the nuclide and element temperature and s(a,b) properties. Then, the order of precedence in setting the temperatures and s(a,b) for nuclides would be:

1) nuclide
2) cell
3) material

However, this could require several modifications to the fortran code in addition to the Python API that I wasn't willing to commit to yet. I'll look into whether this is feasible without a significant rewrite of how s(a,b) and temperatures are handled.

File: openmc/material.py [hunk 1]
```diff
@@ -523,6 +523,95 @@ def remove_element(self, element):
             if element == elm:
                 self._elements.remove(elm)
 
+    def add_material(self, material, percent, percent_type='ao'):
+        """Add a material to the material
```
Comment (wbinventor) [inline openmc/material.py]: This might be less ambiguous if worded as "Mix a material with this material". I would even consider renaming the method as `mix_material(...)` but that's up to you and any others who wish to comment on this.

File: openmc/material.py [hunk 2]
```diff
@@ -523,6 +523,95 @@ def remove_element(self, element):
...
+        percent : float
+            Atom or weight percent
+        percent_type : {'ao', 'wo'}
+            'ao' for atom percent and 'wo' for weight percent
+
+        """
+
+        if self._macroscopic is not None:
+            msg = 'Unable to add a Material to Material ID="{0}" as a ' \
+                  'macroscopic data-set has already been added'.format(self._id)
```
Comment (wbinventor) [inline openmc/material.py]: Consider using `self.id` instead of `self._id` here and elsewhere. I don't think we are perfectly consistent, but it is a goal to use the properties rather than reference attributes directly.

File: openmc/material.py [hunk 3]
```diff
@@ -523,6 +523,95 @@ def remove_element(self, element):
...
+            raise ValueError(msg)
+
+        # Get the atom densities in the given material
+        nuclides_densities = material.get_nuclide_atom_densities()
+        nuclides = nuclides_densities.keys()
+        densities = nuclides_densities.values()
+
+        # Compute the sum of the atom densities
+        sum_densities = 0.
+        for nuclide,density in zip(nuclides,densities):
```
Comment (wbinventor) [inline openmc/material.py]: add a space after the commas

File: openmc/material.py [hunk 4]
```diff
@@ -523,6 +523,95 @@ def remove_element(self, element):
...
+                else:
+                    atomic_mass = openmc.data.atomic_mass(nuclide)
+
+                amm = material.average_molar_mass
+                density *= atomic_mass / amm
+
+            sum_densities += density
+
+        # Add the nuclides of the given material to this material
+        for nuclide,density in zip(nuclides,densities):
```
Comment (wbinventor) [inline openmc/material.py]: add spaces after the commas

File: openmc/material.py [hunk 5]
```diff
@@ -523,6 +523,95 @@ def remove_element(self, element):
...
+
```
Comment (wbinventor) [inline openmc/material.py]: Remove blank line

File: openmc/material.py [hunk 6]
```diff
@@ -523,6 +523,95 @@ def remove_element(self, element):
...
+            # Make a local variable with the nuclide density
+            density = density[1]
+
+            # Convert to weight percent, if requested
```
Comment (wbinventor) [inline openmc/material.py]: What do you think about condensing these two largely redundant loops into a single one that (1) computes `sum_densities` *and* (2) adds the nuclides to the material with the unnormalized densities? A subsequent loop could be used to normalize the densities with `sum_densities`. Perhaps it's not worth it, just something to consider.

File: openmc/material.py [hunk 7]
```diff
@@ -523,6 +523,95 @@ def remove_element(self, element):
...
+                density *= atomic_mass / amm
+
+            # Normalize the density to sum to 1.0
+            density /= sum_densities
+
+            # Add the nuclide to the material
+            self.add_nuclide(nuclide, density*percent, percent_type)
+
+        # Add the s(a,b) tables
+        for sab in material._sab:
```
Comment (wbinventor) [inline openmc/material.py]: Ditto use `material.sab`

File: openmc/material.py (hunk 1 above)
Comment (samuelshaner) [inline openmc/material.py]: I like the changing of the wording for the comment, but I think it is important to keep the method name consistent since it shares so many characteristics to the `add_element` and `add_nuclide` methods.

**Comment by samuelshaner** [comment]
**One question in addition to my comments is what happens when two materials which share the same nuclide are mixed? If I understand correctly, only density of the nuclide in the material parameter will be retained. Perhaps the Material.add_nuclide(...) method could be extended to include an boolean argument increment which will increment a nuclide's density if it already exists in a Material?** 

When two materials that share the same nuclide are mixed, the resulting material will have two nuclide objects of the same nuclide type, but different densities. This would be the same behavior as in the case that someone adds a nuclide to a material twice. For instance:

    fuel1 = openmc.Material()
    fuel1.set_density('g/cm3', 10.0)
    fuel1.add_element('U', 1.0, enrichment=2.5)
    fuel1.add_nuclide('O16', 2.0)
 
    fuel2 = openmc.Material()
    fuel2.set_density('g/cm3', 10.0)
    fuel2.add_element('U', 1.0, enrichment=2.5)
    fuel2.add_nuclide('O16', 1.0)
    fuel2.add_nuclide('O16', 1.0)
 
I believe OpenMC has the same isotopic compositions for fuel1 and fuel2, but the second material object just has two separate nuclides for `O16`.

If we are going to go the route of incrementing the nuclide density, I would opt to just have this done in all cases.

I'll plan on changing the `add_nuclide` method such that the nuclide density is incremented if the nuclide already exists in the material.

File: openmc/material.py (hunk 2 above)
Comment (samuelshaner) [inline openmc/material.py]: I'll change `self._id` to `self.id` throughout

File: openmc/material.py (hunk 6 above)
Comment (samuelshaner) [inline openmc/material.py]: Sure, I'll save the density and then just condense the second loop to fewer lines.

**Comment by samuelshaner** [comment]
I believe all the comments have now been addressed. I addressed the comments in the following manner:

- Ignore duplicate sab tables.
- Increment nuclide density if the nuclide already exists in a material.
- When an sab table is added to a material, check whether another sab table already treats one of the nuclides treated by the given sab table. If two sab tables treat the same nuclide, raise an error.

File: openmc/data/thermal.py [hunk 8]
```diff
@@ -43,6 +43,32 @@
                   'uuo2': 'c_U_in_UO2', 'u-o2': 'c_U_in_UO2', 'u/o2': 'c_U_in_UO2',
                   'zrzrh': 'c_Zr_in_ZrH', 'zr-h': 'c_Zr_in_ZrH', 'zr/h': 'c_Zr_in_ZrH'}
 
+_THERMAL_NUCLIDES = {'c_Al27': ['Al27'],
+                     'c_Be': ['Be9'],
+                     'c_BeO': ['Be9'],
+                     'c_Be_in_BeO': ['Be9'],
+                     'c_Benzine': ['C12', 'H1'],
+                     'c_Ca_in_CaH2': ['Ca40'],
```
Comment (paulromano) [inline openmc/data/thermal.py]: Should this not be all isotopes?

File: openmc/data/thermal.py [hunk 9]
```diff
@@ -43,6 +43,32 @@
...
+                     'c_D_in_D2O': ['H2'],
+                     'c_Fe56': ['Fe56'],
+                     'c_Graphite': ['C12'],
```
Comment (paulromano) [inline openmc/data/thermal.py]: Probably safer to put `['C0', 'C12', 'C13']`

File: openmc/data/thermal.py [hunk 10]
```diff
@@ -43,6 +43,32 @@
...
+                     'c_H_in_CaH2': ['H1'],
+                     'c_H_in_CH2': ['H1'],
+                     'c_H_in_H2O': ['H1'],
+                     'c_H_in_ZrH': ['H1'],
+                     'c_liquid_CH4': ['H1'],
+                     'c_Mg24': ['Mg24'],
+                     'c_O_in_BeO': ['O16'],
```
Comment (paulromano) [inline openmc/data/thermal.py]: How about O17 and O18?

File: openmc/data/thermal.py [hunk 11]
```diff
@@ -43,6 +43,32 @@
...
+                     'c_ortho_D': ['H2'],
+                     'c_ortho_H': ['H1'],
+                     'c_O_in_UO2': ['O16'],
+                     'c_SiO2': ['O16'],
+                     'c_para_D': ['H2'],
+                     'c_para_H': ['H1'],
+                     'c_solid_CH4': ['H1'],
+                     'c_U_in_UO2': ['U238'],
+                     'c_Zr_in_ZrH': ['Zr90']}
```
Comment (paulromano) [inline openmc/data/thermal.py]: Other isotopes of Zr?

File: openmc/data/thermal.py [hunk 12]
```diff
@@ -43,6 +43,32 @@
...
+                     'c_H_in_CH2': ['H1'],
+                     'c_H_in_H2O': ['H1'],
+                     'c_H_in_ZrH': ['H1'],
+                     'c_liquid_CH4': ['H1'],
+                     'c_Mg24': ['Mg24'],
+                     'c_O_in_BeO': ['O16'],
+                     'c_ortho_D': ['H2'],
+                     'c_ortho_H': ['H1'],
+                     'c_O_in_UO2': ['O16'],
+                     'c_SiO2': ['O16'],
```
Comment (paulromano) [inline openmc/data/thermal.py]: How about Si? And other isotopes of O?

File: openmc/data/thermal.py [hunk 13]
```diff
@@ -43,6 +43,32 @@
                   'uuo2': 'c_U_in_UO2', 'u-o2': 'c_U_in_UO2', 'u/o2': 'c_U_in_UO2',
                   'zrzrh': 'c_Zr_in_ZrH', 'zr-h': 'c_Zr_in_ZrH', 'zr/h': 'c_Zr_in_ZrH'}
 
+_THERMAL_NUCLIDES = {'c_Al27': ['Al27'],
+                     'c_Be': ['Be9'],
+                     'c_BeO': ['Be9'],
+                     'c_Be_in_BeO': ['Be9'],
+                     'c_Benzine': ['C12', 'H1'],
```
Comment (paulromano) [inline openmc/data/thermal.py]: Include C0 and C13 to be safe

**Comment by samuelshaner** [comment]
In looking at the sab tables further, there are a few issues with how sab tables are formatted in different xs libraries and how OpenMC treats sab tables that prevent a clean method to check for overlapping sab tables in a material. The issues are:

- Some sab tables in different libraries treat different nuclides. For instance, the `c_Be_in_BeO` table in the ENDF 7.1 library included with MCNP treats Be9 while the same table in the NNDC library treats Be9 and O16.
- There are some overlaps in sab tables that I'm not sure how to interpret. Specifically, the `c_Be_in_BeO` and `c_O_in_BeO` tables in the NNDC library both treat Be9 and O16.
- OpenMC only allows an sab table to be applied to one nuclide. OpenMC will only apply the sab table to the first nuclide found in the material's list of nuclides.

This PR has opened up a bit of a can of worms for how sab tables are to be handled. My inclination is to change the way sab tables are handled such that an sab table is assigned to a **nuclide** or **element** instead of a **material** in input files. Then, the fortran code in OpenMC would need to be modified such that every nuclide is assigned an sab table (in most instances this will be NONE). Thoughts?

**Comment by paulromano** [comment]
Ugh, yeah this is getting messy. So the ENDF thermal scattering files themselves don't really say "use this for these nuclides". That is something that an NJOY user has to specify (ACER, card 8a) which explains why you're seeing differences between different data libraries. As far as I can tell, the fact that NNDC assigns both Be9 and O16 on the "Be in BeO" table is wrong (same with "O in BeO"). There is an [older evaluation](https://www-nds.iaea.org/public/download-endf/ENDF-B-VI-8/THERMAL-SCATTERING/mat/endf-b-vi-8_tsl_27.txt) of BeO that is supposed to be used for both Be and O, but since ENDF/B-VII.1, they've been separated. Probably an issue of someone using an old input with the newer files. The other problem is that NJOY evidently only lets you assign three nuclides. It's not clear at all to me what happens for something like Zr in ZrH where you have 5 naturally-occurring isotopes.

The last issue you mention should be an easy fix regardless of whether we're changing what S(a,b) tables get attached to or not.

I'll have to think more about your proposed change. At present, I lean towards keeping things as they are. Since the information we need to store is the mapping of S(a,b) table -> set of nuclides, it seems appropriate to me that that is done at the level of a material. I would not want to store an attribute that indicates for each nuclide in each material whether it has an S(a,b) table attached (could end up wasting a lot of memory to stores zeros for, say, a large reactor model that is being depleted).

Another alternative is to specifically have moderator nuclides (this approach is used by MC21/NDEX). That is, you could create a material with `<nuclide name="H_in_H2O" ... />` or something like that. I can imagine two different implementations of such a scheme:
1. Explicitly create these nuclides, e.g., add functionality to our `openmc.data` API to create a moderator nuclide. The moderator nuclide would have its own HDF5 data.
2. Have code on the OpenMC Fortran side preprocess data to create these moderator nuclides. In our `nuclides(:)` array, you could have two separate entries for H1, one representing free H1 and the other H1 bound in H2O. They would both contain the same free atom data but one of them would also have S(a,b) data attached.

I don't think there's really a "best" option. They all have pros and cons. I would also think of the decision as being three different pieces:
1. The abstraction presented to the user.
2. The abstraction for how the data is stored on disk.
3. The abstraction for how data is stored in memory within OpenMC.

The three don't necessarily have to all be the same. Keeping the abstraction more in line with how ACE works makes it easier to convert data/inputs and to compare codes on an equal footing. Moving away from the ACE abstraction gives us a little more freedom to represent things as we think they should be.

**Comment by samuelshaner** [comment]
@paulromano - thanks for the detailed explanation! I like the way the laid out the different abstractions to decide on. Based on your comments and my poking around in the source code, my thoughts are as follows:

- **User abstraction**: It seems more natural to have the sab table applied to a **nuclide** or **element** and not a material since an sab table only applies to specific nuclides in a material. This approach also appears less error prone than having sab tables assigned to a material for two reasons: 

1) If two materials are homogenized (e.g. water and ZrH) and the user requests to have sab tables for H in H2O and H in ZrH, this case could be treated just fine with sab tables applied on a nuclide/element basis but it would be unclear how to apply the sab tables if they were assigned to the material. 
2) If the user requests Be_in_BeO and O_in_BeO and uses the NNDC library (or another library where the nuclides are incorrectly assigned to a table), it would be unclear which table to apply to the Be and O isotopes if the table is applied to the material. 

- **Disk storage abstraction**: I'm not too familiar with the `openmc.data` module and feel that the moderator nuclide approach might be foreign to most users (it is to me) so I vote to keep the disk storage abstraction as it stands.
- **Memory storage abstraction**: Good point on not storing a sab table index for each nuclide. The current mapping on sab tables -> nuclides is a one-to-one mapping whereas we really want a one-to-many mapping for sab tables like Zr_in_ZrH where multiple nuclides should be treated by the same table. I propose using the same one-to-one mapping and data structure as is currently implemented, but storing multiple instances of an sab table when it treats more than one nuclide. There are only a few cases where a table treats more than one nuclide, so the extra memory from duplicating sab tables should be trivial.

In summary, I propose:
- Adding the ability for a user to set an sab table on a nuclide or element basis. For backwards compatibility, keep the ability to set sab tables on a material basis, but issue a warning that this capability is deprecated and sab tables set on a nuclide basis will override sab tables set on a materials basis.
- Duplicate sab tables that treat multiple nuclides to fix existing issue with sab tables only mapping to one nuclide.

**Comment by paulromano** [comment]
I'm not too concerned about the NNDC data issue. In fact, I already have a branch with a fix for it (namely, by patching the data when it is converted with `openmc-get-nndc-data`) as well as a fix for applying S(a,b) tables to multiple nuclides as you've suggested. I see your point about the homogenization. However, it sounds like we're operating under the assumption that it is ok to have multiple copies of the same nuclide. Right now, we do not allow duplicate nuclides in tallies because there isn't really any way to disambiguate them. By extension, you couldn't get nuclide-specific MGXSs for these nuclides that are the same but with different S(a,b) tables applied.

**Comment by nelsonag** [comment]
FWIW, I am partial to the moderator-nuclides created in the `openmc.data` package (and automatically created by the `openmc-get-nndc-data` scripts and equivalents).  It seems the clear way through the problem.  It fixes the multiple nuclide problem with the only cost of having to re-learn.  

The implementation of it could be tricky - yes its easy to handle on openmc.data, but we'd need to merge the `sab_header` and `nuclide_header` modules and probably make the S(a,b) scattering be just another angle/energy law.

**Comment by tjlaboss** [comment]
Regarding @nelsonag 's https://github.com/mit-crpg/openmc/pull/773#issuecomment-268220833 :

> If so, we should agree on what we want the behavior to be for dealing with multiple nuclides, and (as you point out), how one could assign an S(a,b) to some of the atoms but not all.

This will be possible with PR [#897: Partial S(alpha, beta) tables](https://github.com/mit-crpg/openmc/pull/897).

**Comment by paulromano** [comment]
There's nothing major stopping this now that we have the ability to have partial S(a,b) tables. Someone just needs to bring it to the finish line.

**Comment by eepeterson** [comment]
I'm interested in this functionality as well as adding a volume_mix class method to Material. I don't want to step on any toes, but I'd be up for trying to finish this if it has just gone stale. How should we go about this?

**Comment by paulromano** [comment]
@eepeterson Glad to hear you're interested in picking up this PR and bringing it to the finish line. Since this PR was originally submitted, a new capability was added in the code to have "partial" S(a,b) tables which provides a straightforward solution to the problem of what to do if you're mixing two materials with different S(a,b) tables applied -- just use the fractions accordingly with `material.add_s_alpha_beta(..., fraction=...)`. Otherwise, the only thing that's needed is essentially a rebase on the current develop branch (or perhaps just a copy/paste of the relevant function), and the addition of a test for the new method(s).

I'm going to go ahead and close this PR, so once you have something ready, please submit a new one.
